Mathlib/Data/Matroid/Closure.lean

```diff
@@ -0,0 +1,199 @@
+    X ⊆ M.cl Y := by
+  rw [← cl_inter_ground]; exact M.subset_cl_of_subset (subset_inter hXY hX)
+
+lemma exists_of_cl_ssubset (hXY : M.cl X ⊂ M.cl Y) : ∃ e ∈ Y, e ∉ M.cl X := by
+  by_contra! hcon
+  exact hXY.not_subset (M.cl_subset_cl_of_subset_cl hcon)
+
+lemma mem_cl_of_mem (M : Matroid α) (h : e ∈ X) (hX : X ⊆ M.E := by aesop_mat) : e ∈ M.cl X :=
+  (M.subset_cl X) h
+
+lemma mem_cl_of_mem' (M : Matroid α) (heX : e ∈ X) (h : e ∈ M.E := by aesop_mat) : e ∈ M.cl X := by
+  rw [← cl_inter_ground]
+  exact M.mem_cl_of_mem ⟨heX, h⟩
+
+lemma not_mem_of_mem_diff_cl (he : e ∈ M.E \ M.cl X) : e ∉ X :=
+  fun heX ↦ he.2 <| M.mem_cl_of_mem' heX he.1
+
+@[aesop unsafe 10% (rule_sets := [Matroid])]
+lemma mem_ground_of_mem_cl (he : e ∈ M.cl X) : e ∈ M.E :=
+  (M.cl_subset_ground _) he
+
+lemma cl_iUnion_cl_eq_cl_iUnion (M : Matroid α) (Xs : ι → Set α) :
+    M.cl (⋃ i, M.cl (Xs i)) = M.cl (⋃ i, Xs i) := by
+  refine (M.cl_subset_cl_of_subset_cl
+    (iUnion_subset (fun i ↦ M.cl_subset_cl (subset_iUnion _ _)))).antisymm ?_
+  rw [← cl_inter_ground, iUnion_inter]
+  refine M.cl_subset_cl (iUnion_subset (fun i ↦ (M.subset_cl _).trans ?_))
+  rw [cl_inter_ground]
+  exact subset_iUnion (fun i ↦ M.cl (Xs i)) i
+
+lemma cl_iUnion_congr (Xs Ys : ι → Set α) (h : ∀ i, M.cl (Xs i) = M.cl (Ys i)) :
+    M.cl (⋃ i, Xs i) = M.cl (⋃ i, Ys i) := by
+  simp [h, ← M.cl_iUnion_cl_eq_cl_iUnion]
+
+lemma cl_biUnion_cl_eq_cl_sUnion (M : Matroid α) (Xs : Set (Set α)) :
+    M.cl (⋃ X ∈ Xs, M.cl X) = M.cl (⋃₀ Xs) := by
+  rw [sUnion_eq_iUnion, biUnion_eq_iUnion, cl_iUnion_cl_eq_cl_iUnion]
+
+lemma cl_biUnion_cl_eq_cl_biUnion (M : Matroid α) (Xs : ι → Set α) (A : Set ι) :
+    M.cl (⋃ i ∈ A, M.cl (Xs i)) = M.cl (⋃ i ∈ A, Xs i) := by
+  rw [biUnion_eq_iUnion, M.cl_iUnion_cl_eq_cl_iUnion, biUnion_eq_iUnion]
+
+lemma cl_biUnion_congr (M : Matroid α) (Xs Ys : ι → Set α) (A : Set ι)
+    (h : ∀ i ∈ A, M.cl (Xs i) = M.cl (Ys i)) : M.cl (⋃ i ∈ A, Xs i) = M.cl (⋃ i ∈ A, Ys i) := by
+  rw [← cl_biUnion_cl_eq_cl_biUnion, iUnion₂_congr h, cl_biUnion_cl_eq_cl_biUnion]
+
+lemma cl_cl_union_cl_eq_cl_union (M : Matroid α) (X Y : Set α) :
+    M.cl (M.cl X ∪ M.cl Y) = M.cl (X ∪ Y) := by
+  rw [eq_comm, union_eq_iUnion, ← cl_iUnion_cl_eq_cl_iUnion, union_eq_iUnion]
+  simp_rw [Bool.cond_eq_ite, apply_ite]
+
+@[simp] lemma cl_union_cl_right_eq (M : Matroid α) (X Y : Set α) :
+    M.cl (X ∪ M.cl Y) = M.cl (X ∪ Y) := by
+  rw [← cl_cl_union_cl_eq_cl_union, cl_cl, cl_cl_union_cl_eq_cl_union]
+
+@[simp] lemma cl_union_cl_left_eq (M : Matroid α) (X Y : Set α) :
+    M.cl (M.cl X ∪ Y) = M.cl (X ∪ Y) := by
+  rw [← cl_cl_union_cl_eq_cl_union, cl_cl, cl_cl_union_cl_eq_cl_union]
+
+@[simp] lemma cl_insert_cl_eq_cl_insert (M : Matroid α) (e : α) (X : Set α) :
+    M.cl (insert e (M.cl X)) = M.cl (insert e X) := by
+  simp_rw [← singleton_union, cl_union_cl_right_eq]
+
+@[simp] lemma cl_union_cl_empty_eq (M : Matroid α) (X : Set α) :
```

Comment:
Is `M.cl ∅ = ∅` true? If so, that seems like a good simp lemma?

Reply:
No - for instance, if `M` is the matroid arising from a function `v : α → V` for a vector space `V`, then `M.cl ∅ = v ⁻¹' {0}`.

Reply:
Thanks! That is a helpful example.